src/libsystemd-network/fuzz-dhcp-server.c

```diff
@@ -17,36 +17,59 @@ ssize_t sendmsg(int sockfd, const struct msghdr *msg, int flags) {
         return 0;
 }
 
-static void add_lease(sd_dhcp_server *server, const struct in_addr *server_address, uint8_t i) {
+static int add_lease(sd_dhcp_server *server, const struct in_addr *server_address, uint8_t i) {
+        _cleanup_(dhcp_lease_freep) DHCPLease *lease = NULL;
         static const uint8_t chaddr[] = {3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3};
-        DHCPLease *lease;
+        int r;
 
         assert(server);
 
-        assert_se(lease = new0(DHCPLease, 1));
+        lease = new0(DHCPLease, 1);
+        if (!lease)
+                return -ENOMEM;
+
+        lease->client_id.data = malloc(2);
+        if (!lease->client_id.data)
+                return -ENOMEM;
+
         lease->client_id.length = 2;
-        assert_se(lease->client_id.data = malloc(2));
         lease->client_id.data[0] = 2;
         lease->client_id.data[1] = i;
+
         lease->address = htobe32(UINT32_C(10) << 24 | i);
         lease->gateway = server_address->s_addr;
         lease->expiration = UINT64_MAX;
         lease->htype = ARPHRD_ETHER;
         lease->hlen = ETH_ALEN;
         memcpy(lease->chaddr, chaddr, ETH_ALEN);
-        assert_se(hashmap_ensure_put(&server->bound_leases_by_client_id, &dhcp_lease_hash_ops, &lease->client_id, lease) >= 0);
-        assert_se(hashmap_ensure_put(&server->bound_leases_by_address, NULL, UINT32_TO_PTR(lease->address), lease) >= 0);
-        lease->server = server;
+
+        lease->server = server; /* This must be set just before hashmap_put(). */
+
+        r = hashmap_ensure_put(&server->bound_leases_by_client_id, &dhcp_lease_hash_ops, &lease->client_id, lease);
+        if (r < 0)
+                return r;
+
+        r = hashmap_ensure_put(&server->bound_leases_by_address, NULL, UINT32_TO_PTR(lease->address), lease);
+        if (r < 0)
+                return r;
```

Comment:
hmm, doesn't this add a new "bad" failure code path?

previously, oom issues here were covered by assert(). now you turned those into regular error checks, which is fine, but if the first hashmap_ensure_put() but the second one fails, then you'll have the lease "half-added"...

i.e. it would be cleaner if the second failure path would undo the first insertion again